### ISSUE TITLE
New Generics

### ISSUE THREAD

**Issue by peq** [wurstscript/WurstScript#679]
Currently, generics are translated using erasure. This means, that generics can only be instantiated with types compatible with `int`. The idea of this proposal is to change the translation and instead of using erasure translation would generate one copy per instantiated type. This would allow to extend generics with some interesting new features, but would also be incompatible in some cases.

Language changes:
============

## 1. Allow all types for generics

Every type can be inserted for a generic type parameter without any restrictions.

For example, we can then use `LinkedList<vec3>`.

This also allows us to get rid of the implicit calls to `fromIndex` and `toIndex` that we currently have as a woraround.

## 2. Allow to restrict type parameters / type classes

Sometimes we want to restrict type parameters.
For example `LinkedList<T>.toString() returns string` only can be implemented, if T has a `toString` method.

This is comparable to [statically resolved type parameters in F#](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/generics/statically-resolved-type-parameters), type classes in Haskell and a bit like type bounds in Java.

Syntactically this can look somewhat like this:

```
typeclass Show
    function toString() returns string

function LinkedList<T>.toString<T : Show>() returns string
```

```
typeclass Serializable
    function encode() returns string
    static function decode(string s) returns thistype

function LinkedList<T>.encode<T : Serializable>() returns string
function LinkedList<T>.decode<T : Serializable>() returns string
```

Type classes would be instantiated implicitly (it is enough to implement the required methods). We would also extend extension functions so that new static methods can be implemented for a type.
If there are multiple fitting methods in scope, it is a compiletime error. The instantiation is part of the type, so the following example is not allowed:

```
package A

public typeclass Ord
    function lessEq(thistype other) returns boolean

function string.lessEq(string other) returns boolean
    return this.length() <= other.length()

public class TreeSet<T : Ord>
    ...

public TreeSet<string> mySet = TreeSet.create("a","aa","aaa")

package B
import A

function string.lessEq(string other) returns boolean
    return /* some other order */

init
    TreeSet<string> s = mySet // error, because Ord intance / lessEq is not compatible


```

## 3. Disallow casting to int

Since we now allow all types to be used for type parameters, we can no longer cast type parameters to `int`. However, we can use the new feature of restricted type parameters to implement the same feature:

Old `HashMap`:

```
public class HashMap<K,V> extends Table

	/** Whether a value exists under the given key or not */
	function has(K key) returns boolean
		return hasInt(key castTo int)

	/** Saves the given value under the given key */
	function put(K key, V value)
		saveInt(key castTo int, value castTo int)

	/** Retrieves the value saved under the given key */
	function get(K key) returns V
		return loadInt(key castTo int) castTo V

	/** Removes the value saved under the given key */
	function remove(K key)
		removeInt(key castTo int)
```

New `HashMap`

```
typeclass Indexable
    function toIndex() returns int
    static function fromIndex() returns thistype

public class HashMap<K : Indexable,V : Indexable> extends Table

	/** Whether a value exists under the given key or not */
	function has(K key) returns boolean
		return hasInt(key.toIndex())

	/** Saves the given value under the given key */
	function put(K key, V value)
		saveInt(key.toIndex(), value.toIndex())

	/** Retrieves the value saved under the given key */
	function get(K key) returns V
		return V.fromIndex(loadInt(key.toIndex()))

	/** Removes the value saved under the given key */
	function remove(K key)
		removeInt(key.toIndex())
```

To simplify the transition, we could automatically interpret the old code as the new one.

## 4. Disallowing casting between different instantiations

Currently the following code is valid:

```
class A
class B extends A

function foo(LinkedList<A> l)

@Test function testCast()
	LinkedList<B> list = asList(new B, new B, new B)
	foo(list castTo int castTo LinkedList<A>)
```

With the proposed changes this would still compile, but might no longer work, since translation might create complete different code for `List<A>` and `List<B>`. 
Maybe it makes sense to guarantee that all instantiations with class types share the same code (same as it is now), but this is difficult to implement in combination with typeclasses.

**Comment by Cokemonkey11** [comment]
> The idea of this proposal is to change the translation and instead of using erasure translation would generate one copy per instantiated type.

Can you explain what this means/how this works? I'm interested to know if the compiled jass looks any different.

It seems that this is a pretty breaking change, and one that less experienced programmers are more likely to be confused by. vJass, older C++, and older Java (including what you might see as a student/junior engineer) prefer inheritance over type interfacing.

Is it possible to ease the burden on wurst users, for example by gating this behind a feature flag?

```
# wurst.build
wurst_generics = "2019"
```

**Comment by Frotty** [comment]
> Can you explain what this means/how this works? I'm interested to know if the compiled jass looks any different.

Right now generic classes are simply classes, and the generic type is cast to int via the typecasting bug. Thus it is erased during compilation, similar to Java.

The plan is to actually keep the type (allowing us to use the actual respective hashtable functions, e.g.) and not casting it to int anymore.
This will of course mean that if you have a LinkedList<Indexable> and LinkedList<MyInterface>, it will result in 2 classes in the compiled jass.

> It seems that this is a pretty breaking change, [...]

Not necessarily. Peq already mentioned that the old code could still be interpreted the old way, and basic usage of generics will stay compatible. Only unsafe casting like `LinkedList<X> castTo LinkedList<Y>' would break.

>  Is it possible to ease the burden on wurst users, for example by gating this behind a feature flag?

Sure, but I don't think it is ever useful to make language features dependent on passed arguments. It will just always be a problem/confusion in the future, I believe.

**Comment by Cokemonkey11** [comment]
> Not necessarily. Peq already mentioned that the old code could still be interpreted the old way, and basic usage of generics will stay compatible. Only unsafe casting like LinkedList<X> castTo LinkedList<Y> would break.

Is that also true for casting up/down an inheritance hierarchy?

That's the use-case I envisage newer users to be abusing

**Comment by peq** [comment]
> Is that also true for casting up/down an inheritance hierarchy?

No, you would still be able to cast `A` to `B` if `A` and `B` have a common super-type. Only if `A` and `B` are unrelated types (like `LinkedList<Unit>` and `LinkedList<Entity>`) there would be incompatibilities. And these cases already give you a compilation error, so the only way you can notice the incompatibility is if you cast to int first.

So I think this would be a minor incompatibility.

**Comment by peq** [comment]
Updated section 2 on "restrict type parameters"

**Comment by peq** [comment]
Updated section 2 and again. With these changes interfaces stay like they are, which should avoid some confusion for users and compiler implementers. 

I'll try to release a first version on the weekend (See PR #931).
Then we'll try to update the standard library and probably will find more problems.

**Comment by AndyClausen** [comment]
> inb4 WurstScript becomes Typescript

Thank you all for still working on this. It's making me want to give map-making a shot. I was almost scared away when I today looked up JASS, but seeing this makes me ~~hard~~ excited. Although, I might wait until this is ready for user testing.

Keep up the good work! And don't let Reforged kill your spirits ❤️

**Comment by Frotty** [comment]
Thanks. If by `Although, I might wait until this is ready for user testing` you mean waiting for this feature to be implemented before trying out Wurstscript, I don't think that's a good idea 😅 .
Reforged itself isn't so bad I guess, but killing classic and blizzard's general handling of wc3/the situation discouraged me somewhat and I have been inactive for some time.

**Comment by Cokemonkey11** [comment]
@peq revive new generics? :)

**Comment by peq** [comment]
Just started a real job this year. Probably gonna be some time before I'll be hacking Wurst again.

**Comment by Frotty** [comment]
The typeclass proposal was never completed, but the practical part is now covered by the current generics implementation: generic specializations lower directly to Jass/Lua without the old integer typecasting. With optimizations, an `ArrayList` get/set can reduce to a direct array access. Closing this RFC as superseded; typeclasses can be proposed separately if needed.
